File: algorithms/leet.0115.src.1.java
public class Solution {
    public int numDistinct(String S, String T) {
        int K = S.length(), L = T.length();
        int[] C = new int[L+1];
        C[0] = 1;
        for(int i = 0; i < K; ++i) {
            for(int j = L; j > 0; --j) {
                if(T.charAt(j-1) == S.charAt(i)) {
                    C[j] += C[j-1];
                }
            }
        }
        return C[L];
    }
}

File: algorithms/leet.0115.src.1.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        n = len(s)
        m = len(t)
        dp = [0] * (m + 1)
        dp[0] = 1
        for i, c in enumerate(s):
            for j in range(m, 0, -1):
                if t[j-1] != c:
                    continue
                dp[j] += dp[j-1]
            # print(dp)
        return dp[m]
